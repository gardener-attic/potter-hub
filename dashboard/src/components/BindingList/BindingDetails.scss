.BindingDetails code {
  word-wrap: break-word;
}
