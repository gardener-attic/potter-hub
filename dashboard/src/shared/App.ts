import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { HELM_URL } from "./Kube";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = HELM_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, keepHistory: boolean) {
    let keepHistoryQuery;
    if (keepHistory) {
      keepHistoryQuery = "keepHistory=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, keepHistoryQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
