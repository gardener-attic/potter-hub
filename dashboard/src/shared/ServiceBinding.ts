import { axiosWithAuth } from "./AxiosInstance";
import { K8S_REVERSE_PROXY_URL } from "./Kube";
import { definedNamespaces } from "./Namespace";
import { ICondition, ServiceCatalog } from "./ServiceCatalog";

interface IK8sApiSecretResponse {
  kind: string;
  apiVersion: string;
  metadata: {
    selfLink: string;
    resourceVersion: string;
  };
  data: { [s: string]: string };
}

interface IServiceBinding {
  metadata: {
    name: string;
    selfLink: string;
    uid: string;
    resourceVersion: string;
    creationTimestamp: string;
    finalizers: string[];
    generation: number;
    namespace: string;
  };
  spec: {
    externalID: string;
    instanceRef: {
      name: string;
    };
    secretName: string;
  };
  status: {
    conditions: ICondition[];
    asyncOpInProgress: boolean;
    currentOperation: string;
    reconciledGeneration: number;
    operationStartTime: string;
    externalProperties: {};
    orphanMitigationInProgress: boolean;
    unbindStatus: string;
  };
}

export interface IServiceBindingWithSecret {
  binding: IServiceBinding;
  secret?: IK8sApiSecretResponse;
}

export class ServiceBinding {
  public static async create(
    bindingName: string,
    instanceRefName: string,
    namespace: string,
    parameters: {},
  ) {
    const url = ServiceBinding.getLink(namespace);
    const { data } = await axiosWithAuth.post<IServiceBinding>(url, {
      metadata: {
        name: bindingName,
      },
      spec: {
        instanceRef: {
          name: instanceRefName,
        },
        parameters,
      },
    });
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const url = this.getLink(namespace, name);
    return axiosWithAuth.delete(url);
  }

  public static async get(namespace: string, name: string) {
    const url = this.getLink(namespace, name);
    const { data } = await axiosWithAuth.get<IServiceBinding>(url);
    return data;
  }

  public static async list(namespace?: string): Promise<IServiceBindingWithSecret[]> {
    const bindings = await ServiceCatalog.getItems<IServiceBinding>("servicebindings", namespace);

    return Promise.all(
      bindings.map(binding => {
        const { secretName } = binding.spec;
        const ns = binding.metadata.namespace;
        return axiosWithAuth
          .get<IK8sApiSecretResponse>(this.secretEndpoint(ns) + secretName)
          .then(response => {
            return { binding, secret: response.data };
          })
          .catch(err => {
            // return with undefined secrets
            return { binding };
          });
      }),
    );
  }

  private static getLink(namespace?: string, name?: string): string {
    return `${K8S_REVERSE_PROXY_URL}/apis/servicecatalog.k8s.io/v1beta1${
      namespace ? `/namespaces/${namespace}` : ""
    }/servicebindings${name ? `/${name}` : ""}`;
  }

  private static secretEndpoint(namespace: string = definedNamespaces.default): string {
    return `${K8S_REVERSE_PROXY_URL}/api/v1/namespaces/${namespace}/secrets/`;
  }
}
