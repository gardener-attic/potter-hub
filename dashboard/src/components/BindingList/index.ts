import BindingList from "./BindingList";

export default BindingList;
