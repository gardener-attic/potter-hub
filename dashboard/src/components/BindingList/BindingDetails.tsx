import * as React from "react";

import TerminalModal from "../../components/TerminalModal";
import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";
import "./BindingDetails.css";

interface IBindingDetailsState {
  modalIsOpen: boolean;
}

class BindingDetails extends React.Component<IServiceBindingWithSecret, IBindingDetailsState> {
  public state = {
    modalIsOpen: false,
  };

  public render() {
    const { binding, secret } = this.props;
    const { secretName } = binding.spec;

    let secretDataArray: string[][] = [];
    if (secret) {
      const secretData = Object.keys(secret.data).map(k => [k, atob(secret.data[k])]);
      secretDataArray = [...secretData];
    }
    let message = "";
    if (secretDataArray.length > 0) {
      message = "";
      secretDataArray.forEach(statusPair => {
        const [key, value] = statusPair;
        message += `${key}: ${value}\n`;
      });
    } else {
      message = "The secret is empty";
    }
    return (
      <dl className="BindingDetails container margin-normal">
        <dt key={secretName}>
          {secretName} <a onClick={this.openModal}>(show)</a>
        </dt>
        <TerminalModal
          modalIsOpen={this.state.modalIsOpen}
          closeModal={this.closeModal}
          title={`Secret: ${secretName}`}
          message={message}
        />
      </dl>
    );
  }

  public openModal = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };
}

export default BindingDetails;
