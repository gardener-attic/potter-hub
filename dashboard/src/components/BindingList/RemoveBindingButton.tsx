import * as React from "react";

import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";

interface IRemoveBindingButtonProps {
  bindingWithSecret: IServiceBindingWithSecret;
  removeBinding: (name: string, ns: string) => Promise<boolean>;
}

class RemoveBindingButton extends React.Component<IRemoveBindingButtonProps> {
  public render() {
    return (
      <div className="RemoveBindingButton">
        <button
          className="button button-small button-danger"
          onClick={this.handleRemoveBindingClick}
        >
          Remove
        </button>
      </div>
    );
  }

  private handleRemoveBindingClick = async () => {
    const {
      removeBinding,
      bindingWithSecret: { binding },
    } = this.props;
    const { name, namespace } = binding.metadata;
    removeBinding(name, namespace);
  };
}

export default RemoveBindingButton;
