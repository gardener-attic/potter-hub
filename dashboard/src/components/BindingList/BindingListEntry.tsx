import * as React from "react";
import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";

import TerminalModal from "../TerminalModal";
import BindingDetails from "./BindingDetails";
import RemoveBindingButton from "./RemoveBindingButton";

import "./BindingList.css";

interface IBindingListEntryProps {
  bindingWithSecret: IServiceBindingWithSecret;
  removeBinding: (name: string, namespace: string) => Promise<boolean>;
}

interface IBindingListEntryState {
  modalIsOpen: boolean;
}

class BindingListEntry extends React.Component<IBindingListEntryProps, IBindingListEntryState> {
  public state = {
    modalIsOpen: false,
  };

  public render() {
    const {
      bindingWithSecret,
      bindingWithSecret: { binding },
    } = this.props;
    const { name } = binding.metadata;

    let reason = <span />;
    let message = "";
    const condition = [...binding.status.conditions]
      .sort((a, b) => a.lastTransitionTime.localeCompare(b.lastTransitionTime))
      .pop();
    if (condition) {
      reason = <code>{condition.reason}</code>;
      message = condition.message;
    }

    return (
      <tr>
        <td>{name}</td>
        <td>{reason}</td>
        <td>
          <button className="button button-small" onClick={this.openModal}>
            Show message
          </button>
          <TerminalModal
            modalIsOpen={this.state.modalIsOpen}
            closeModal={this.closeModal}
            title="Status Message"
            message={message}
          />
        </td>
        <td>
          <BindingDetails {...bindingWithSecret} />
        </td>
        <td>
          <RemoveBindingButton {...this.props} />
        </td>
      </tr>
    );
  }

  public openModal = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };
}

export default BindingListEntry;
