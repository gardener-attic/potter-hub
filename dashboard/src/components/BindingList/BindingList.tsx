import * as React from "react";

import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";
import BindingListEntry from "./BindingListEntry";

import "./BindingList.css";

interface IBindingList {
  bindingsWithSecrets: IServiceBindingWithSecret[];
  removeBinding: (name: string, namespace: string) => Promise<boolean>;
}

class BindingList extends React.Component<IBindingList> {
  public render() {
    const { removeBinding, bindingsWithSecrets } = this.props;
    return (
      <div className="BindingList">
        <table>
          <thead>
            <tr>
              <th>Binding</th>
              <th>Status</th>
              <th>Message</th>
              <th>Secret</th>
              <th />
            </tr>
          </thead>
          <tbody>
            {bindingsWithSecrets.length > 0 ? (
              bindingsWithSecrets.map(b => [
                <BindingListEntry
                  key={b.binding.metadata.uid}
                  bindingWithSecret={b}
                  removeBinding={removeBinding}
                />,
              ])
            ) : (
              <tr>
                <td> No bindings found</td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    );
  }
}

export default BindingList;
