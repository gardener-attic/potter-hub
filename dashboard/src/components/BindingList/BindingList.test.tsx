import { shallow } from "enzyme";
import * as React from "react";

import { IServiceBindingWithSecret } from "shared/ServiceBinding";
import BindingList from "./BindingList";

it("renders information about the binding with a secret", () => {
  const bindingWithSecret = {
    binding: {
      metadata: {
        name: "foo",
        namespace: "default",
        selfLink: "",
        uid: "",
        resourceVersion: "",
        creationTimestamp: "",
        finalizers: [],
        generation: 1,
      },
      spec: {
        externalID: "",
        instanceRef: {
          name: "foo",
        },
        secretName: "bar",
      },
      status: {
        asyncOpInProgress: false,
        currentOperation: "",
        reconciledGeneration: 1,
        operationStartTime: "",
        externalProperties: {},
        orphanMitigationInProgress: false,
        unbindStatus: "",
        conditions: [
          {
            lastTransitionTime: "1",
            type: "a type",
            status: "good",
            reason: "none",
            message: "everything okay here",
          },
        ],
      },
    },
    secret: {
      kind: "",
      apiVersion: "",
      metadata: {
        name: "bar",
        namespace: "default",
        selfLink: "",
        uid: "",
        resourceVersion: "",
        creationTimestamp: "",
        finalizers: [],
        generation: 1,
      },
      data: {
        mySecret: "Y29udGVudAo=",
      },
    },
  } as IServiceBindingWithSecret;
  const wrapper = shallow(
    <BindingList bindingsWithSecrets={[bindingWithSecret]} removeBinding={jest.fn()} />,
  );
  expect(wrapper).toMatchSnapshot();
});
