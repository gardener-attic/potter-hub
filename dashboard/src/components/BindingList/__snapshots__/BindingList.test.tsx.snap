// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders information about the binding with a secret 1`] = `
<div
  className="BindingList"
>
  <table>
    <thead>
      <tr>
        <th>
          Binding
        </th>
        <th>
          Status
        </th>
        <th>
          Message
        </th>
        <th>
          Secret
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <BindingListEntry
        bindingWithSecret={
          Object {
            "binding": Object {
              "metadata": Object {
                "creationTimestamp": "",
                "finalizers": Array [],
                "generation": 1,
                "name": "foo",
                "namespace": "default",
                "resourceVersion": "",
                "selfLink": "",
                "uid": "",
              },
              "spec": Object {
                "externalID": "",
                "instanceRef": Object {
                  "name": "foo",
                },
                "secretName": "bar",
              },
              "status": Object {
                "asyncOpInProgress": false,
                "conditions": Array [
                  Object {
                    "lastTransitionTime": "1",
                    "message": "everything okay here",
                    "reason": "none",
                    "status": "good",
                    "type": "a type",
                  },
                ],
                "currentOperation": "",
                "externalProperties": Object {},
                "operationStartTime": "",
                "orphanMitigationInProgress": false,
                "reconciledGeneration": 1,
                "unbindStatus": "",
              },
            },
            "secret": Object {
              "apiVersion": "",
              "data": Object {
                "mySecret": "Y29udGVudAo=",
              },
              "kind": "",
              "metadata": Object {
                "creationTimestamp": "",
                "finalizers": Array [],
                "generation": 1,
                "name": "bar",
                "namespace": "default",
                "resourceVersion": "",
                "selfLink": "",
                "uid": "",
              },
            },
          }
        }
        key=""
        removeBinding={[MockFunction]}
      />
    </tbody>
  </table>
</div>
`;
